server/src/main/java/org/elasticsearch/reservedstate/service/ReservedClusterStateService.java

```diff
@@ -170,6 +170,8 @@ public void onFailure(Exception e) {
                         if (isNewError(existingMetadata, reservedStateVersion.version())) {
                             logger.debug("Failed to apply reserved cluster state", e);
                             errorListener.accept(e);
+                        } else {
+                            errorListener.accept(null);
```

Comment:
This is the actual bug, we never called the callback to ensure the async watcher is released.